cairo-vm-tracer/static/tracer.js

```diff
@@ -0,0 +1,387 @@
+
+  var fp = entry.fp;
+  for (var i = 0; i < MAX_STACK_TRACE && fp != initial_fp; i++) {
+    const pc = memory[fp - 1];
+    stack_trace.append(create_stack_trace_row(fp, pc));
+    $("#mem_row" + fp).css("border-top", "2px black solid");
+    fp = memory[fp - 2];
+  }
+}
+
+function create_stack_trace_row(fp, pc) {
+  const fp_cell = $("<td>").append(fp);
+  const pc_cell = $("<td>").append(pc);
+  const row = $("<tr>").append(fp_cell).append(pc_cell);
+
+  pc_cell.mouseenter(function () {
+    $(".inst" + pc).addClass("highlight_instruction");
+  });
+  pc_cell.mouseleave(function () {
+    $(".inst" + pc).removeClass("highlight_instruction");
+  });
+
+  return row;
+}
+
+function create_watch_row() {
+  const watch_expr = $("<input>").attr({ type: "text" });
+  const watch_result = $("<span>");
+
+  watch_exprs.push({ watch_expr: watch_expr, watch_result: watch_result });
+
+  watch_expr.keyup(function (event) {
+    if (event.key == "Enter") {
+      update_watch();
+    }
+  });
+
+  const n_watches = watch_exprs.length;
+  watch_expr.change(function (event) {
+    // Add new watch if needed.
+    if (watch_expr.val() != "" && watch_exprs.length == n_watches) {
+      $("#watch_table").append(create_watch_row());
+    }
+
+    update_watch();
+  });
+
+  // Make sure navigation keys (s, n, ...) will not work when watch_expr is focused.
+  watch_expr.keypress(function (event) {
+    event.stopPropagation();
+  });
+
+  return $("<tr>")
+    .append($("<td>").append(watch_expr))
+    .append($("<td>").append(watch_result));
+}
+
+var update_watch_ajax = { abort: function () {} };
+
+function update_watch() {
+  // REMOVE RETURN TO EVALUATION EXPRESSION
+  return;
+  var query_str = "eval.json?step=" + encodeURIComponent(current_step);
+  for (const entry of watch_exprs) {
+    const expr_txt = entry.watch_expr.val();
+    query_str +=
+      "&expr=" + encodeURIComponent(expr_txt == "" ? "null" : expr_txt);
+  }
+  // Abort previous AJAX request if exists.
+  update_watch_ajax.abort();
+  update_watch_ajax = $.getJSON(query_str, function (data) {
+    for (var idx = 0; idx < data.length; ++idx) {
+      watch_exprs[idx].watch_result.text(data[idx]);
+    }
+  });
+}
+
+/*
+  Clears the text selection in the window.
+  This function was copied from
+  https://stackoverflow.com/questions/880512/prevent-text-selection-after-double-click.
+*/
+function clearSelection() {
+  if (document.selection && document.selection.empty) {
+    document.selection.empty();
+  } else if (window.getSelection) {
+    var sel = window.getSelection();
+    sel.removeAllRanges();
+  }
+}
+
+function toggle_breakpoint(event) {
+  $(this).toggleClass("breakpoint");
+  clearSelection();
+  event.stopPropagation();
+}
+
+function has_breakpoint(step) {
+  return (
+    $(".inst" + trace[step].pc).hasClass("breakpoint") ||
+    $("#mem_row" + memory_accesses[step].dst).hasClass("breakpoint") ||
+    $("#mem_row" + memory_accesses[step].op0).hasClass("breakpoint") ||
```

Comment:
Shouldn't the `".inst"` be `"#mem_row"` like the others? Breakpoint functionality seems not to be working, and making this change seems to fix it.

Reply:
Thanks for catching this! It's fixed now